### DISCUSSION TITLE
Reduce 함수 활용에 관한 질문

### DISCUSSION THREAD

**Discussion by unuhqueen** [DeveloperAcademy-POSTECH/MacC-GetARock#127] [Q&A]
이번 주디의 코드리뷰 덕분에 for문을 Reduce 함수로 바꾸는 대장정을 거쳤는데요,,, 27b43a220c0adcd8eecb6c6936b71172256f3f54
Reduce 함수로 바꾸면서 엥 이건 왜 안되지?가 궁금한 부분이 있어서 여쭤봅니다.

참고로 기존 함수는 다음과 같습니다.
```swift
private func calculateNumberOfBandMemberText() -> Int {
        var sum = 0
        for num in 0...selectedBand.filledPosition.count - 1 {
            sum += selectedBand.filledPosition[num].numberOfPerson
        }
        return sum
}
```

제가 Reduce로 바꾼 함수는 다음과 같은데요
```swift
private lazy var numberOfBandMember: Int = selectedBand.filledPosition.reduce(0) { $0 + $1.numberOfPerson }
```

이 함수의 클로저를 축약하지 않은 원형으로 바꾸면 다음과 같습니다.
```swift
selectedBand.filledPosition.reduce(0) { (partial: Int, current: Band.PositionSet) -> Int in
            return partial + current.numberOfPerson
}
```
여기서 궁금한 점은 왜 이런식으로 코드를 짜면 에러가 나고 위에처럼 짜면 에러가 안나는지 궁금했습니다..!!
```swift
selectedBand.filledPosition.reduce(0) { (partial: Int, current: Band.PositionSet.numberOfPerson) -> Int in
            return partial + current
}
```

에러 내용은 다음과 같은데요, 여기서 왜 이 context에서는 `Band.PositionSet` 안에 있는 `numberOfPerson`을 사용할 수 없다고 하는지 궁금했습니다.

<img width="626" alt="스크린샷 2023-01-05 오전 11 12 57" src="https://user-images.githubusercontent.com/49133245/210686025-54e3b2c7-08e0-41ba-81d2-1c84dbb1118f.png">

읽어주셔서 감사합니다!

**Comment by EuniceNam** [comment]
Band.PositionSet은 array라서 Band.PositionSet[인덱스].numberOfPerson이 아니면 접근할 수 없는 것 같아요. 

리듀스에서 인수를 Element로 받을 뿐이라서, 리듀스와 관계없이 없는 식별자라 못 쓰는 것 같아요.
